Update the Microsoft.IdentityModel.* and Microsoft.Identity.Web.* packages to the latest patch release (7.0.3 & 2.15.2)

mkArtakMSFT opened the pull request:
# Update the Microsoft.IdentityModel.* and Microsoft.Identity.Web.* packages to the latest patch release (7.0.3 & 2.15.2)

Update the reference to the Microsoft.IdentityModel.* and Microsoft.Identity.Web.* packages so that we don't regress AAD authentication scenarios for web apps.

## Description

We've hit an issue with AAD authentication in ASP.NET Core web apps, which was resulting in errors during login. This was due to an issue in the IdentityModel package, for which @halter73 has proposed a fix: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2361
The Identity team has approved the fix and has released a new package to NuGet so that we can update our dependency and **avoid the regression in 8.0**.

Please note, that this change will have to include the soure-build change as well: https://github.com/dotnet/source-build-externals/pull/228

Fixes https://github.com/dotnet/aspnetcore/issues/51005

## Customer Impact

Customers who will try to use AAD authentication for their ASP.NET Core web applications in 8.0 will fail to login.

## Regression?

- [x] Yes
- [ ] No

This was technically an existing bug, which was already in the IdentityModel package, however only after a recent change https://github.com/dotnet/aspnetcore/pull/49542 the issue has surfaced impacting 8.0 apps.
 
## Risk

- [ ] High
- [ ] Medium
- [x] Low

From our point of view this is a dependency update. And the dependency has taken only a targeted fix to avoid the bug, going through all the necessary validation on the AAD side.

## Verification

- [x] Manual (required)
- [ ] Automated

## Packaging changes reviewed?

- [ ] Yes
- [ ] No
- [x] N/A

----

## When servicing release/2.1

- [ ] Make necessary changes in eng/PatchConfig.props

mkArtakMSFT committed:
Update IdentityModel to v 7.0.3

We've made a fix to the IdentityModel validation logic, which has introduced a regression in .NET 8 AAD authentication scenarios. The AAD team has approved the fix and have release a new package to NuGet. This change updates our reference to the patched version, so that we avoid the regression impacting 8.0 apps.

halter73 commented on eng/Versions.props:
```diff
@@ -12,7 +12,7 @@
     <PreReleaseVersionIteration>
     </PreReleaseVersionIteration>
     <ValidateBaseline>true</ValidateBaseline>
-    <IdentityModelVersion>7.0.0</IdentityModelVersion>
```
I don't think this will be enough for our templates to pick up my `AadIssuerValidator` fix from https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2361.

The fix is now published as part of [Microsoft.IdentityModel.Validators version 7.0.3](https://www.nuget.org/packages/Microsoft.IdentityModel.Validators/), but none of our package references that use `IdentityModelVersion`, like [Microsoft.IdentityModel.Protocols.OpenIdConnect](https://www.nuget.org/packages/Microsoft.IdentityModel.Protocols.OpenIdConnect) have a dependency on Microsoft.IdentityModel.Validators.

[Microsoft.Identity.Web](https://www.nuget.org/packages/Microsoft.Identity.Web) which is not a dependency of any packages that we ship, but is a dependency of some of our project templates does have a dependency on Microsoft.IdentityModel.Validators, but the latest version of Microsoft.Identity.Web is currently 2.15.1, and it only requires Microsoft.IdentityModel.Validators 6.33.0 which is not recent enough.

@jennyf19 I think we're going to need to update the dependencies for https://www.nuget.org/packages/Microsoft.Identity.Web, publish it to NuGet.org, and then update or dependency versions for those here:

https://github.com/dotnet/aspnetcore/blob/e3aba084cde27f8e718b188e3e9f7239e63d5816/eng/Versions.props#L301-L304

mkArtakMSFT replied:
Good catch, thanks @halter73. Sounds like we will need to wait for a set of new packages to include in this update. 
@eerhardt FYI, so that you delay your change before this is finalized.

jennyf19 replied:
ack. We will get started on this.

jennyf19 replied:
@mkArtakMSFT the release in Id Web (2.15.2) with 7.0.3 IdentityModel is on NuGet
https://github.com/AzureAD/microsoft-identity-web/releases/tag/2.15.2

mkArtakMSFT committed:
Updated Microsoft.Identity.Web.* package versions

mkArtakMSFT commented:
@halter73 I've now included the MIcrosoft.Identity.Web.* packages too and this should be ready now.

mkArtakMSFT commented:
I believe the CI failures will resolve after https://github.com/dotnet/source-build-externals/pull/228 is merged.

halter73 commented on eng/Versions.props:
```diff
@@ -298,10 +298,10 @@
-    <MicrosoftIdentityWebGraphServiceClientVersion>2.13.0</MicrosoftIdentityWebGraphServiceClientVersion>
-    <MicrosoftIdentityWebUIVersion>2.13.0</MicrosoftIdentityWebUIVersion>
-    <MicrosoftIdentityWebDownstreamApiVersion>2.13.0</MicrosoftIdentityWebDownstreamApiVersion>
+    <MicrosoftIdentityWebVersion>2.15.2</MicrosoftIdentityWebVersion>
```
It still looks like this has a 6.33.0 rather than a 7.0.3 dependency on Microsoft.IdentityModel.Validators for everything but the net8.0 TFM.

https://www.nuget.org/packages/Microsoft.Identity.Web/2.15.2#dependencies-body-tab

This is probably okay given we only really need Microsoft.IdentityModel.Validators 7.0.3 for .NET 8 as the old OIDC handler avoids the issue, but it is odd to leave it at 6.33.0 for all the other TFMs. Is 7.0.3 known to be broken on the non-net8.0 TFMs? @jennyf19

halter73 replied:
> I has 7.0.3 for the .NET 8.0 target framework, and 6.33.0 for the other target framework

I noticed that. That's why I mentioned it was "for everything but the net8.0 TFM."

It's not a huge problem. It probably even makes sense for Microsft.AspNetCore.* and Microsoft.Extensions.* where using 7.x packages on necoreapp3.1 to avoid issues related to breaking changes in the ASP.NET Core shared runtime. For example Microsoft.AspNetCore.Authentication.JwtBearer 3.1.x only directly targets necoreapp3.1 and isn't tested on newer runtimes.
 
However, I don't think it makes sense for any of the other packages. You ship Microsoft.IdentityModel.Validators yourself, it has no direct or indirect Microsft.AspNetCore.* or Microsoft.Extensions.* dependency I see, and it directly targets net8.0, net6.0, netstardard2.0, etc.. Why wouldn't you use the latest Microsoft.IdentityModel.Validators as a dependency in the latest Microsoft.Identity.Web net6.0 TFM? The latest Microsoft.IdentityModel.Validators also directly targets net6.0!

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/source-build-externals build 20231018.1

Microsoft.SourceBuild.Intermediate.source-build-externals
 From Version 8.0.0-alpha.1.23502.1 -> To Version 8.0.0-alpha.1.23518.1

eerhardt commented:
@halter73 @mkArtakMSFT - The source build leg should now be fixed in this PR. I've updated to include the changes from https://github.com/dotnet/source-build-externals/pull/229.

halter73 commented:
@dotnet/aspnet-build @dotnet/dnceng I posted about this on our internal "ASP.NET Build" teams chat, but I'll ask here too for wider visibility.

We are having an issue updating our project templates to reference the latest version of [Microsoft.Identity.Web 2.15.2](https://www.nuget.org/packages/Microsoft.Identity.Web/2.15.2#dependencies-body-tab) because it depends on [Microsoft.Identity.Web.TokenAcquisition 2.15.2](https://www.nuget.org/packages/Microsoft.Identity.Web.TokenAcquisition/2.15.2#dependencies-body-tab) which in turn depends on RC2 packages that we ship like Microsoft.AspNetCore.Authentication.OpenIdConnect (>= 8.0.0-rc.2.23480.2).
 
The problem is that our template tests see our Microsoft.AspNetCore.Authentication.OpenIdConnect 8.0.0-ci artifacts as a forced downgrade and fails our template tests with a build error like:
 
```
/datadisks/disk1/work/BCE909F9/w/B40D0985/e/Templates/BaseFolder/AspNet.vxrmciskazme/AspNet.vxrmciskazme.csproj : error NU1605: Warning As Error: Detected package downgrade: Microsoft.AspNetCore.Authentication.JwtBearer from 8.0.0-rc.2.23480.2 to 8.0.0-ci. Reference the package directly from the project to select a different version.
/datadisks/disk1/work/BCE909F9/w/B40D0985/e/Templates/BaseFolder/AspNet.vxrmciskazme/AspNet.vxrmciskazme.csproj : error NU1605:  AspNet.vxrmciskazme -> Microsoft.Identity.Web 2.15.2 -> Microsoft.Identity.Web.TokenAcquisition 2.15.2 -> Microsoft.AspNetCore.Authentication.JwtBearer (>= 8.0.0-rc.2.23480.2)
/datadisks/disk1/work/BCE909F9/w/B40D0985/e/Templates/BaseFolder/AspNet.vxrmciskazme/AspNet.vxrmciskazme.csproj : error NU1605:  AspNet.vxrmciskazme -> Microsoft.AspNetCore.Authentication.JwtBearer (>= 8.0.0-ci)
```

https://dev.azure.com/dnceng-public/public/_build/results?buildId=442668&view=ms.vss-test-web.build-test-results-tab 

Have we ever referenced an external package in our project templates that reference a prerelease package we built? I assume not, or else we would have run into this issue before. What should we do to fix this?

wtgodbe commented:
It's probably fine to include `NU1605` in `NoWarn` for those tests - In the bits we actually ship, our templates will depend on a newer `Microsoft.AspNetCore.Authentication.JwtBearer` than the transitive dependency, so it shouldn't cause any real issues. At least some of our templates already NoWarn the reference to JwtBearer: https://github.com/dotnet/aspnetcore/blob/149d04bfd79425cbb297e67ac03b08a75ec8eb12/src/ProjectTemplates/Web.ProjectTemplates/RazorPagesWeb-CSharp.csproj.in#L26

halter73 committed:
Add  NoWarn="NU1605" to WebApi-CSharp.csproj.in

- This is already there in RazorPagesWeb-CSharp.csproj.in and StarterWeb-CSharp.csproj.in

halter73 commented:
Thanks for the suggestion. It looks like we already had been using `NoWarn="NU1605"` for these exact dependencies in RazorPagesWeb-CSharp.csproj.in and StarterWeb-CSharp.csproj.in but not WebApi-CSharp.csproj.in.

jaliyaudagedara commented:
Hi Guys,

Unfortunately, I am experiencing the [issue: 51005](https://github.com/dotnet/aspnetcore/issues/51005) after updating to .NET 8.
```
Bearer was not authenticated. Failure message: IDX40001: Issuer: 'https://something.b2clogin.com/<tenantId>/v2.0/', does not match any of the valid issuers provided for this application.
```
I am using `Microsoft.Identity.Web: 2.15.5`
```
services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
    .AddMicrosoftIdentityWebApi(Configuration, "AzureAdB2C");
```

I could get around by allowing all the issuers, but that's not good.
```
services.Configure<JwtBearerOptions>(JwtBearerDefaults.AuthenticationScheme, options =>
{
    options.TokenValidationParameters.IssuerValidator = 
        (string issuer, SecurityToken securityToken, TokenValidationParameters validationParameters) => issuer;
});
```
